robottelo/report_portal/portal.py

```diff
@@ -0,0 +1,257 @@
+
+    def _test_params(self, status, defect_type):
+        """Customise parameters for Test items API request
+
+        :returns dict: The parameters dict for API test items request
+        """
+        params = [
+            ('filter.eq.launch', self.info['id']),
+            ('page.page', 1),
+            ('page.size', 50),
+            ('page.sort', 'start_time'),
+        ]
+        rp_defect_types = ReportPortal.defect_types
+        if defect_type:
+            if defect_type in rp_defect_types:
+                params.insert(0, ('filter.eq.issue$issue_type', rp_defect_types[defect_type]))
+            else:
+                raise ValueError(
+                    f'Invalid value \'{defect_type}\' for defect type parameter, '
+                    f'should be one of {[*rp_defect_types.keys()]}'
+                )
+        rp_statuses = ReportPortal.statuses
+        if status:
+            if status in rp_statuses:
+                params.insert(2, ('filter.eq.status', status))
+            else:
+                raise ValueError(
+                    f'Invalid value \'{status}\' for status parameter, '
+                    f'should be one of {rp_statuses}'
+                )
+        return dict(params)
+
+    @retry(
+        stop=stop_after_attempt(4), wait=wait_fixed(10),
+    )
+    def _test_requester(self, params, page):
+        """The Test Items GET requester to fetch the data on a page
+
+        If any error and before Failing explicitly, it retries for 3 times with 10 seconds delay
+
+        :returns tuple (int, list): Total pages count and the list of tests along with
+            each tests properties in a page
+        """
+        params['page.page'] = page
+        resp = requests.get(
+            url=f'{self.report_portal.api_url}/item',
+            headers=self.report_portal.headers,
+            params=params,
+            verify=False,
+        )
+        resp.raise_for_status()
+        total_pages = resp.json()['page']['totalPages']
+        pagedata = resp.json().get('content')
+        return total_pages, pagedata
+
+    def tests(self, status=None, defect_type=None):
+        """Returns tests data customized by kwargs parameters.
+
+        This is a main function that will be called to retrieve the tests data
+        of a particular test status or/and defect_type
+
+        :param str status: Filter tests of a launch with tests `status`
+        :param str defect_type: Filter tests of a launch with tests `defect_type`
+        :returns dict: All filtered tests dict based on params data along with each
+            tests properties
+        """
+        page = 1
+        params = self._test_params(status, defect_type)
+        total_pages, data = self._test_requester(params=params, page=page)
+        while page < total_pages:
+            page += 1
+            _, pagedata = self._test_requester(params=params, page=page)
+            data.extend(pagedata)
+        # returning only test names
```

Comment:
Optional:

I know within the collection we only need the test names when we pass a status, but I think this function would be far more versatile if it returns more data for the tests. That way the caller can apply whatever  additional logic that they want.

We could generate a dictionary, keyed on the test name, so the caller can use `keys()`.

Reply:
True! And earlier I was returning all the data for a test. But it was our(me&U) meeting which decides to return only test names.

Never mind I can improve in that regards.